typing/typecore.ml

```diff
@@ -4183,18 +4388,19 @@ and type_expect_
         match expected_mode.position with
         | RNontail ->
           raise (Error (loc, env, Exclave_in_nontail_position))
-        | RTail (mode, _) ->
+        | RTail (regionality, _) ->
           (* mode' is RNontail, because currently our language cannot construct
              region in the tail of another region.*)
-          let mode' = mode_default mode in
+          let mode' = mode_default {expected_mode.mode with locality = regionality} in
           (* The middle-end relies on all functions which allocate into their
              parent's region having a return mode of local. *)
-          submode ~loc ~env ~reason:Other Value_mode.local mode';
+          submode ~loc ~env ~reason:Other {mode'.mode with locality = Mode.Regionality.local} mode';
           let new_env = Env.add_exclave_lock env in
           let exp =
             type_expect ?in_function ~recarg new_env mode' sbody ty_expected_explained
           in
-          submode ~loc ~env ~reason:Other Value_mode.regional expected_mode;
+          submode ~loc ~env ~reason:Other {expected_mode.mode
+            with locality = Mode.Regionality.regional} expected_mode;
```

Comment:
this submode (and the one above) looks ugly right now. They will be better with `min_with_locality` in https://github.com/ocaml-flambda/ocaml-jst/pull/160.